Libs/Optimize/ParticleSystem/VtkMeshWrapper.cpp

```diff
@@ -0,0 +1,822 @@
+      // When at the intersection of two faces while also being at the edge of the mesh, the edge-sliding will keep alternating
+      // between the two faces without actually going anywhere since it is at a corner in the mesh.
+      //std::cerr << "exiting due to face repetition\n";
+      break;
+    }
+
+    if (facesTraversed.size() > 1000) {
+      std::cerr << "Warning, more than 1000 faces traversed\n";
+      for (int i = 0; i < facesTraversed.size(); i++) {
+        std::cerr << facesTraversed[i] << ", ";
+      }
+      //std::cerr << "\nPositions: ";
+      //for (int i = 0; i < positions.size(); i++) {
+      //    std::cerr << "(" << positions[i][0] << "," << positions[i][1] << "," << positions[i][2]
+      //                << ") ";
+      //  }
+      /*
+        std::cerr << "\n\n";
+        std::cerr << "ID: " << idx << "\n";
+        std::cerr << "Original point: " << PrintValue<Eigen::Vector3d>(point_a) << "\n";
+        std::cerr << "Original vector: " << PrintValue<Eigen::Vector3d>(projected_vector) << "\n";
+        std::cerr << "Current point: " << PrintValue<Eigen::Vector3d>(currentPoint) << "\n";
+        std::cerr << "remaining vector: " << PrintValue<Eigen::Vector3d>(remainingVector) << "\n";
+        std::cerr << "currentBary: " << PrintValue<vec3>(currentBary) << "\n";
+        std::cerr << "targetPoint: " << PrintValue<Eigen::Vector3d>(targetPoint) << "\n";
+        std::cerr << "targetBary: " << PrintValue<vec3>(targetBary) << "\n";
+        std::cerr << std::endl;
+        */
+      break;
+
+    }
+
+    if (targetBary[0] + barycentricEpsilon >= 0 && targetBary[1] + barycentricEpsilon >= 0 &&
+        targetBary[2] + barycentricEpsilon >= 0 && targetBary[0] - barycentricEpsilon <= 1 &&
+        targetBary[1] - barycentricEpsilon <= 1 && targetBary[2] - barycentricEpsilon <= 1) {
+      currentPoint = targetPoint;
+      break;
+    }
+
+    std::vector<int> negativeVertices;
+    for (int i = 0; i < 3; i++) {
+      if (targetBary[i] < 0) {
+        negativeVertices.push_back(i);
+      }
+    }
+
+    if (negativeVertices.size() == 0 || negativeVertices.size() > 2) {
+      std::cerr << "ERROR: invalid number of negative vertices. Point is not on surface.\n";
+      break;
+    }
+    int negativeEdge = negativeVertices[0];
+    Eigen::Vector3d intersect = GetBarycentricIntersection(currentBary, targetBary, currentFace,
+                                                           negativeEdge);
+
+    // When more than 1 negative barycentric coordinate, compute both intersections and take the closest one.
+    if (negativeVertices.size() == 2) {
+      int negativeEdge1 = negativeVertices[1];
+      Eigen::Vector3d intersect1 = GetBarycentricIntersection(currentBary, targetBary, currentFace,
+                                                              negativeEdge1);
+
+      double length0 = (intersect - currentPoint).norm();
+      double length1 = (intersect1 - currentPoint).norm();
```

Comment:
(may not matter) these can be switched to squared norm to avoid the square root.

Reply:
Will change.